### DISCUSSION TITLE
DevToys.Extensions.FluentIconFinder

### DISCUSSION THREAD

**Discussion by pierre3** [DevToys-app/DevToys#1399] [Show and tell]
Thank you for the wonderful tool!

I am currently trying to create an extension for DevToys.
Since it was difficult to find the glyphs of FluentSystemIcons available by default, I created an extension that allows for easy searching.

https://github.com/pierre3/DevToys.Extensions.FluentIconFinder
I would appreciate any feedback.

I would like to share some issues I noticed during implementation.

1. The timing of the DataGrid’s OnRowSelected event seems to be off.
    - No response on the first click.
    - On the second click, the information of the row clicked on the first click is retrieved.
    - Additionally, after the first click, if I click on an area outside the DataGrid, the event is triggered.
2. Since the glyph that can be specified in the Icon is of type char, surrogate pair icons cannot be specified.

**Comment by veler** [comment]
Hi,
This is an awesome idea and well executed! Thank you for your feedback on the DataGrid.

If that's OK, I will add a link to this extension in the DevToys documentation. Here is how it looks on my local machine currently.

![image](https://github.com/user-attachments/assets/69074ce9-faf3-47e1-ad01-9affa2aac109)

**Reply by pierre3** [reply]
Thanks for considering adding the link to the extension in the DevToys documentation. I’d really appreciate it!

**Reply by pierre3** [reply]
Currently, we are considering a method to display surrogate pair icons as well. 

It seems that we can handle this by changing all Glyph properties from char to int and modifying the assignment to data-glyph in [FontIcon.razor](https://github.com/DevToys-app/DevToys/blob/4bbb9d159dcc3a7993a109bafb2e1f3e7f570019/src/app/dev/DevToys.Blazor/Components/Feedback/FontIcon/FontIcon.razor#L26) with the following code:

```cs
data-glyph=@(char.ConvertFromUtf32(Glyph))
```

However, applying this change will also alter the interface of the extensions, causing existing extensions to stop working.

Another option I am considering is adding a new UIIcon (UIIcon32) that supports surrogate pair icons.

https://github.com/pierre3/DevToys/commit/197e3f3e4cf1621458e080b32e6e17aa975a7f19#diff-dabb397b135469a892dd53cf98cb3da494ddffd106d95704455c5da603b8c552R26

With this implementation, existing extensions will continue to work. However, feature icons and existing components with embedded icons will still only support Glyphs up to 16-bit.

**Reply by veler** [reply]
Thank you :)

I have a follow-up question: do you have an example of case where surrogate characters should be used?

Also, I'm wondering if we could detect the surrogate character automatically (using `char.IsSurrogate`) and adjust the behavior of `FontIcon/UIIcon` when it detects one, instead of creating a new UIIcon32.

What do you think?

**Reply by pierre3** [reply]
Thank you for your reply. 

Currently, Fluent UI Icon has about 8,100 icons, but approximately 1,600 of them have Glyph values that exceed the 2-byte limit (65535). As a result, these 1,600 icons cannot be used with the current implementation. 

FluentSystemIcons-Regular.json
https://github.com/microsoft/fluentui-system-icons/blob/38a6f9837c2d45018678cb38668ebff83763fe58/fonts/FluentSystemIcons-Regular.json#L6398

![image](https://github.com/user-attachments/assets/1f9c149e-c896-4717-a4fd-af24220c9146)

Additionally, since the remaining information is lost when passed as a char type argument, it would be difficult to detect and adjust surrogate characters in the internal implementation.

**Reply by veler** [reply]
Ah, yes, that makes sense! And indeed, we can't use `char` because we may lose some information. I see that in your proposal you use `int` instead of `char`, which makes sense.
I'm wondering now if, from API perspective, we could just do 2 endpoints like this:

```csharp
public static partial class GUI
{
    public static IUIIcon Icon(string fontName, char glyph)
    {
        return Icon(id: null, fontName, glyph);
    }

    // Note how the method name doesn't mention `32`.
    public static IUIIcon32 Icon(string fontName, int glyph)
    {
        return Icon32(id: null, fontName, glyph);
    }
}
```

That could potentially make it more smooth for extension developers. Also, perhaps `IUIIcon32` could inherits from `IUIIcon` or vice versa, whatever's simpler.

**Reply by pierre3** [reply]
Thank you for the suggestion! Your proposed implementation is indeed more elegant. I’ll try making the changes along with implementing the inheritance from IUIIcon. Once the implementation is done, I’ll create a Pull Request, so please wait for it.

**Reply by veler** [reply]
Thank you very much! Appreciate :D

**Reply by jerone** [reply]
> That could potentially make it more smooth for extension developers.

This will indeed make it backwards-compatible for existing extensions.

**Comment by pierre3** [comment]
@veler 
I have created a pull request. Please review the changes.

#1435
